### ISSUE TITLE
Segmentation fault on Cumulus Linux devices running agent

### ISSUE THREAD

**Issue by LukoJy3D** [prometheus/prometheus#12200]
### What did you do?

No specific actions were done. Prometheus agents are running fine for a few months and then crash out of the blue

```
VERSION="Cumulus Linux 3.7.15"
ID_LIKE=debian

CPU: 2 core Intel(R) Atom(TM) CPU  C2338  @ 1.74GHz
Memory: 4GB
```

### What did you see instead? Under which circumstances?

Service crashes, reinstall does not work, and no other service on the device is affected.

### System information

Linux 4.1.0-cl-7-amd64 x86_64

### Prometheus version

```text
Can't get the version when running prometheus --version
Segmentation fault (core dumped)
```

### Prometheus configuration file

```yaml
#
# Ansible managed
#

global:
  scrape_interval: 15s
  scrape_timeout: 10s

remote_write:
  - URL: long term storage endpoint
    queue_config:
      max_shards: 10

scrape_configs:
  - job_name: exporters
    file_sd_configs:
    - files:
      - /etc/prometheus/targets/*.yml
    relabel_configs:
    - target_label: instance
      replacement: yyy.zzz.com
    - target_label: roles
      replacement: network
  - job_name: prometheus
    static_configs:
    - targets:
      - localhost:9090
    relabel_configs:
    - target_label: instance
      replacement: yyy.zzz.com
```

### Logs

```text
2023-03-30T11:43:59.905799+00:00 yyy.zzz.com systemd[1]: Unit prometheus.service entered failed state.
2023-03-30T11:44:05.136870+00:00 yyy.zzz.com systemd[1]: prometheus.service holdoff time over, scheduling restart.
2023-03-30T11:44:05.191063+00:00 yyy.zzz.com kernel: [35920114.412416] prometheus[31963]: segfault at 31b6da4 ip 000000000045747c sp 00007ffc06feacd0 error 7 in prometheus[400000+2283000]
2023-03-30T11:44:05.192500+00:00 yyy.zzz.com systemd[1]: prometheus.service: main process exited, code=killed, status=11/SEGV
2023-03-30T11:44:05.195823+00:00 yyy.zzz.com systemd[1]: Unit prometheus.service entered failed state.
2023-03-30T11:44:10.390194+00:00 yyy.zzz.com systemd[1]: prometheus.service holdoff time over, scheduling restart.
2023-03-30T11:44:10.431064+00:00 yyy.zzz.com kernel: [35920119.651430] prometheus[31981]: segfault at 31b6da4 ip 000000000045747c sp 00007fff4fe40f40 error 7 in prometheus[400000+2283000]
2023-03-30T11:44:10.435824+00:00 yyy.zzz.com systemd[1]: prometheus.service: main process exited, code=killed, status=11/SEGV
2023-03-30T11:44:10.437235+00:00 yyy.zzz.com systemd[1]: Unit prometheus.service entered failed state.
2023-03-30T11:44:15.636892+00:00 yyy.zzz.com systemd[1]: prometheus.service holdoff time over, scheduling restart.
2023-03-30T11:44:15.652065+00:00 yyy.zzz.com kernel: [35920124.871022] prometheus[31993]: segfault at 31b6da4 ip 000000000045747c sp 00007ffd0db49f20 error 7 in prometheus[400000+2283000]
2023-03-30T11:44:15.652990+00:00 yyy.zzz.com systemd[1]: prometheus.service: main process exited, code=killed, status=11/SEGV
2023-03-30T11:44:15.653843+00:00 yyy.zzz.com systemd[1]: Unit prometheus.service entered failed state.
2023-03-30T11:44:20.886882+00:00 yyy.zzz.com systemd[1]: prometheus.service holdoff time over, scheduling restart.
2023-03-30T11:44:20.907071+00:00 yyy.zzz.com kernel: [35920130.124907] prometheus[32000]: segfault at 31b6da4 ip 000000000045747c sp 00007ffed807fc60 error 7 in prometheus[400000+2283000]
2023-03-30T11:44:20.910242+00:00 yyy.zzz.com systemd[1]: prometheus.service: main process exited, code=killed, status=11/SEGV
2023-03-30T11:44:20.911150+00:00 yyy.zzz.com systemd[1]: Unit prometheus.service entered failed state.
2023-03-30T11:44:26.073294+00:00 yyy.zzz.com systemd[1]: prometheus.service holdoff time over, scheduling restart.
2023-03-30T11:44:26.121895+00:00 yyy.zzz.com systemd[1]: prometheus.service: main process exited, code=killed, status=11/SEGV
```

**Comment by dgl** [comment]
It is strange that prometheus --version crashes. Are there any extra details in dmesg?

I suspect this may be a hardware issue or your Linux distribution not applying microcode updates; I'm not sure the Prometheus project can help you here, you likely need to talk to your distribution to see if they have microcode updates or your hardware vendor to confirm for example that this issue isn't related to https://www.theregister.com/2017/02/06/cisco_intel_decline_to_link_product_warning_to_faulty_chip/

**Comment by dgl** [comment]
Not much to go on there, the fact it's in a core piece of Go code points more to this being not really a Prometheus problem.

One idea that occurs is maybe the code cached in memory has got corrupted. You could try echo 1 > /proc/sys/vm/drop_caches on a bad machine (no guarantees, that might tank performance for a bit as you'll lose all page cache).

**Comment by LukoJy3D** [comment]
Big thanks for your input. Indeed we managed to hit multiple issues on multiple nodes simultaneously. What a timing... 

The first issue is related to hardware, as after the issue was opened, we managed to catch other services throwing segfaults as well.

Secondly, as you pointed out, we reviewed our ansible preflight tasks, and apparently, we did not checksum correctly. Here is a good example if someone messes up with this as well: https://github.com/prometheus-community/ansible/blob/3bc2c9452fa4b42574a7baaeb9bc9796f3df805e/roles/prometheus/tasks/preflight.yml#L91
